pool/web/handler/admin_experiments_assignments.ml

```diff
@@ -22,65 +22,142 @@ let index req =
+               in
+               Lwt_result.return (session, assignments))
+             sessions
+           ||> CCList.all_ok
+           >|+ Page.Admin.Assignment.list experiment context
+         | true ->
+           Lwt_list.fold_left_s
+             (fun res session ->
+               res
+               |> Lwt_result.lift
+               >>= fun sessions ->
+               Assignment.find_deleted_by_session
+                 database_label
+                 session.Session.id
+               >|+ function
+               | [] -> sessions
+               | assignments -> sessions @ [ session, assignments ])
+             (Ok [])
+             sessions
+           >|+ Page.Admin.Assignment.marked_as_deleted experiment context
        in
-       Page.Admin.Assignment.list assignments experiment context
-       |> create_layout req context
-       >|+ Sihl.Web.Response.of_html
+       html |> create_layout req context >|+ Sihl.Web.Response.of_html
   in
   result |> HttpUtils.extract_happy_path req
 ;;
 
+let index req = list req
+let deleted req = list ~marked_as_deleted:true req
+
+let ids_and_redirect_from_req req =
+  let open Pool_common in
+  let experiment_id, session_id, assignment_id =
+    let open Message.Field in
+    HttpUtils.(
+      ( find_id Experiment.Id.of_string Experiment req
+      , find_id Pool_common.Id.of_string Session req
+      , find_id Assignment.Id.of_string Assignment req ))
+  in
+  let%lwt urlencoded = Sihl.Web.Request.to_urlencoded req in
+  let redirect =
+    let open Page.Admin.Assignment in
+    let open CCOption in
+    let experiment_path =
+      Format.asprintf
+        "/admin/experiments/%s"
+        (experiment_id |> Experiment.Id.value)
+    in
+    let to_path =
+      let path = Format.asprintf "%s/%s" experiment_path in
+      function
+      | Assignments -> path "assignments"
+      | DeletedAssignments -> path "assignments/deleted"
+      | Session ->
+        Format.asprintf "sessions/%s" (session_id |> Id.value) |> path
+    in
+    CCList.assoc ~eq:CCString.equal Message.Field.(show Redirect) urlencoded
+    |> CCList.head_opt
+    >>= Page.Admin.Assignment.read_assignment_redirect
+    >|= to_path
+    |> value ~default:experiment_path
+  in
+  Lwt.return (session_id, assignment_id, redirect)
+;;
+
 let cancel req =
   let open Utils.Lwt_result.Infix in
-  let experiment_id, id =
-    let open Pool_common.Message.Field in
-    ( HttpUtils.find_id Experiment.Id.of_string Experiment req
-    , HttpUtils.find_id Assignment.Id.of_string Assignment req )
-  in
-  let redirect_path =
-    Format.asprintf
-      "/admin/experiments/%s/assignments"
-      (Experiment.Id.value experiment_id)
+  let%lwt _, assignment_id, redirect_path = ids_and_redirect_from_req req in
+  let result { Pool_context.database_label; _ } =
+    Utils.Lwt_result.map_error (fun err -> err, redirect_path)
+    @@ let* assignment = Assignment.find database_label assignment_id in
```

Comment:
the only change here is the source of the variable "assignment_id" and "redirect_path". Could you change back the rest, which has just another indentation?